pkg/arvo/sys/vane/clay.hoon

```diff
@@ -1962,32 +1968,163 @@
+            ::  we don't want to ~& an entire dome
+            %-  ~(run by sto.fiz)
+            |=(v=(unit dome:clay) ?~(v %not-received %recieved))
+        ==
+    =.  fiz  (make-melt bas con)
+    (emil moves)
+  ::
+  ++  take-fuse
+    |=  [bec=beak =riot]
+    ^+  ..take-fuse
+    ?~  riot
+      ~&  [%fuse-for syd %missing bec]
+      ::  by setting fiz to *melt the merge is aborted - any further
+      ::  responses we get for the merge will cause take-fuse to crash
+      =.  fiz  *melt
+      ..take-fuse
+    ?>  (~(has by sto.fiz) bec)
+    =.  fiz
+        :+  bas.fiz  con.fiz
+        (~(put by sto.fiz) bec `!<(dome:clay q.r.u.riot))
+    =/  all-done=flag
+        %-  ~(all by sto.fiz)
+        |=  res=(unit dome:clay)
+        ^-  flag
+        !=(res ~)
+    ?.  all-done
+      ..take-fuse
+    =|  rag=rang
+    =/  clean-hut-ran  hut.ran
+    =/  initial-dome=dome:clay  (need (~(got by sto.fiz) bas.fiz))
+    =/  continuation-yaki=yaki
+        (~(got by hut.ran) (~(got by hit.initial-dome) let.initial-dome))
+    =/  parents=(list tako)  ~[(~(got by hit.initial-dome) let.initial-dome)]
+    =/  merges  con.fiz
+    |-
+    ^+  ..take-fuse
+    ?~  merges
+      ::  reset fiz before potentially parking - we hope to
+      ::  succeed so we don't need this state anymore.
+      =.  fiz  *melt
+      =.  hut.ran  clean-hut-ran
+      (park | [%| continuation-yaki(p (flop parents))] rag)
+    =/  [bec=beak g=germ]  i.merges
+    =/  ali-dom=dome:clay  (need (~(got by sto.fiz) bec))
+    =/  result  (merge-helper p.bec q.bec g ali-dom `continuation-yaki)
+    ?-  -.result
+        %|
+      ::  merge failed
+      (mean %fuse-merge-failed p.result)
+    ::
+        %&
+      ::  now we have a unit merge-result
+      =/  merge-result=(unit merge-result)  +.result
+      ?~  merge-result
+        ::  this merge was a no-op, just continue
+        $(merges t.merges)
+      ?~  conflicts.u.merge-result
+        ::  no merge conflicts
+        =/  merged-yaki=yaki
+            ?-  -.new.u.merge-result
+                %|
+              +.new.u.merge-result
+            ::
+                %&
+              ::  convert the yuki to yaki
+              =/  yuk=yuki  +.new.u.merge-result
+              =/  lobes=(map path lobe)
+                  %-  ~(run by q.yuk)
+                  |=  val=(each page lobe)
+                  ^-  lobe
+                  ?-  -.val
+                    %&  (page-to-lobe +.val)
+                    %|  +.val
+                  ==
+              (make-yaki p.yuk lobes now)
+            ==
+        %=  $
+          continuation-yaki  merged-yaki
+          merges  t.merges
+          hut.ran  (~(put by hut.ran) r.merged-yaki merged-yaki)
+          lat.rag  (~(uni by lat.rag) lat.u.merge-result)
+          parents  [(~(got by hit.ali-dom) let.ali-dom) parents]
+        ==
+      :: if there are merge conflicts send the error and abort the merge
+      =.  fiz  *melt
+      (done %& conflicts.u.merge-result)
```

Comment:
Need to restore `hut.ran` at least so we don't leave garbage around.  Though, I don't like the idea of having to specifically restore changed state, since it would be easy to later change more state and not realize it needs to be restored.

Maybe next to `clean-hut-ran` we should pin a clean `..take-fuse` as well and just restore that.

Reply:
In all cases I think we want to restore `hut.ran` and set `fiz` to `*melt`. I guess we could pin `..take-fuse` and then change `fiz` but if we're doing that I think we might as well be explicit.

IMO since we don't know what state is changing/how it should be restored I prefer explicitly restoring those parts. that should be restored and changing those parts that should be changed.

As an aside - I don't think there's a simple way to define a "cleanup handler" (closest analog I can think of) here to execute these state changes but if there is that would also be a way to reduce errors of this kind.

Reply:
I'm definitely fond of "copy-on-write root as contiguous a chunk of per-fallible-operation state as possible", not sure how feasible this is for merging maps or w/e.

Reply:
ok took a shot at this. Given how I implemented this change I had to call park after calling done, which does line up with merge but is different from @philipcmonk's earlier suggestion.